From 0364a50d722e9e3348d08d9d6f4ab5e82d4ce479 Mon Sep 17 00:00:00 2001
From: Gleb Popov <6yearold@gmail.com>
Date: Thu, 28 Aug 2025 14:11:26 +0300
Subject: [PATCH 03/10] sha512.c: Tweaks for Elbrus2000 architecture

---
 crypto/sha/sha512.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/crypto/sha/sha512.c b/crypto/sha/sha512.c
index 4c3a3b6c09..4e89f9b68a 100644
--- a/crypto/sha/sha512.c
+++ b/crypto/sha/sha512.c
@@ -59,7 +59,7 @@
 #include "internal/cryptlib.h"
 #include "crypto/sha.h"
 
-#if defined(__i386) || defined(__i386__) || defined(_M_IX86) || defined(__x86_64) || defined(_M_AMD64) || defined(_M_X64) || defined(__s390__) || defined(__s390x__) || defined(__aarch64__) || defined(SHA512_ASM)
+#if defined(__i386) || defined(__i386__) || defined(_M_IX86) || defined(__x86_64) || defined(_M_AMD64) || defined(_M_X64) || defined(__s390__) || defined(__s390x__) || defined(__aarch64__) || defined(__e2k__) || defined(SHA512_ASM)
 #define SHA512_BLOCK_CAN_MANAGE_UNALIGNED_DATA
 #endif
 
@@ -445,6 +445,11 @@ static const SHA_LONG64 K512[80] = {
                         asm ("rev8 %0, %1"  \
                         : "=r"(ret)         \
                         : "r"(x)); ret; })
+#elif defined(__e2k__)
+#include <x86gprintrin.h>
+#define PULL64(x) __builtin_bswap64(x)
+#define ROTR(x,s)  ( (s) > 48 ? __rolq((x), 64 - (s)) \
+                              : __rorq((x),      (s)) )
 #endif
 #if defined(__riscv_zknh) && __riscv_xlen == 32
 #define Sigma0(x) ({ SHA_LONG64 ret; unsigned int *r = (unsigned int *)(&(ret));    \
@@ -528,6 +533,9 @@ static const SHA_LONG64 K512[80] = {
                         : "=r"(ret)                                 \
                         : "r"(x^z), "r"(y), "r"(x)); ret; })
 #endif
+#elif defined(__e2k__) && __iset__ >= 5
+#define sigma0(x)  (__builtin_e2k_plog (0x96, ROTR((x),1),  ROTR((x),8),  ((x)>>7)))
+#define sigma1(x)  (__builtin_e2k_plog (0x96, ROTR((x),19), ROTR((x),61), ((x)>>6)))
 #elif defined(_MSC_VER)
 #if defined(_WIN64) /* applies to both IA-64 and AMD64 */
 #pragma intrinsic(_rotr64)
-- 
2.50.1

