diff --git a/psx/psx_calls.c b/psx/psx_calls.c
index aff105f..1860595 100644
--- a/psx/psx_calls.c
+++ b/psx/psx_calls.c
@@ -50,7 +50,7 @@
     || defined(__alpha__) || defined(__hppa__) || defined(__sh__) \
     || defined(__m68k__) || defined(__sparc__) || defined(__arc__) \
     || defined(__microblaze__) || defined(__or1k__) \
-    || defined(__xtensa__)
+    || defined(__xtensa__) || defined(__e2k__)
 
 #undef _NSIG
 #undef _NSIG_BPW
@@ -71,7 +71,7 @@
 
 #if defined(__x86_64__) || defined(__i386__) \
     || defined(__arm__) || defined(__powerpc__) \
-    || defined(__arc__)
+    || defined(__arc__) || defined(__e2k__)
 /* field used */
 #define SA_RESTORER  0x04000000
 #endif /* architectures that use SA_RESTORER */
@@ -80,7 +80,7 @@
     || defined(__aarch64__) \
     || defined(__m68k__) || defined(__sh__) \
     || defined(__s390__) || defined(__sparc__) \
-    || defined(__xtensa__)
+    || defined(__xtensa__) || defined(__e2k__)
 /* field defined */
 #define _HAS_SA_RESTORER   void *sa_restorer;
 #else
@@ -206,6 +206,13 @@ static void psx_posix_syscall_actor(int signum, siginfo_t *info, void *ignore) {
 	    __asm__ __volatile__("\npsx_restorer:\n\tli 0, 172\n\tsc\n");
 #elif defined(__arc__)
 	    __asm__ __volatile__("\npsx_restorer:\n\tmov r8, 139\n\ttrap_s 0\n");
+#elif defined(__e2k__) && defined(__ptr64__)
+            __asm__ __volatile__("\npsx_restorer:\n"
+		"{\n\t"
+			"adds,s 0x0, 0xad, %b[0]\n\t"
+                	"sdisp %ctpr1, 0x3\n"
+                "}\n"
+                "call %ctpr1, wbs=0x5\n");
 #else
 #error "unsupported architecture - https://bugzilla.kernel.org/show_bug.cgi?id=219687"
 #endif /* supported architectures */
