diff -Naur a/tests/16-sim-arch_basic.c b/tests/16-sim-arch_basic.c
--- a/tests/16-sim-arch_basic.c	2025-02-06 15:21:01.368787456 +0300
+++ b/tests/16-sim-arch_basic.c	2025-02-06 15:46:08.530429489 +0300
@@ -80,6 +80,9 @@
 	rc = seccomp_arch_add(ctx, SCMP_ARCH_AARCH64);
 	if (rc != 0)
 		goto out;
+	rc = seccomp_arch_add(ctx, SCMP_ARCH_E2K);
+	if (rc != 0)
+		goto out;
 	rc = seccomp_arch_add(ctx, SCMP_ARCH_MIPSEL);
 	if (rc != 0)
 		goto out;
@@ -147,6 +150,9 @@
 	rc = seccomp_arch_remove(ctx, SCMP_ARCH_AARCH64);
 	if (rc != 0)
 		goto out;
+	rc = seccomp_arch_remove(ctx, SCMP_ARCH_E2K);
+	if (rc != 0)
+		goto out;
 	rc = seccomp_arch_remove(ctx, SCMP_ARCH_MIPSEL);
 	if (rc != 0)
 		goto out;
diff -Naur a/tests/23-sim-arch_all_le_basic.c b/tests/23-sim-arch_all_le_basic.c
--- a/tests/23-sim-arch_all_le_basic.c	2025-02-06 15:21:24.233236829 +0300
+++ b/tests/23-sim-arch_all_le_basic.c	2025-02-06 15:46:42.971106956 +0300
@@ -59,6 +59,9 @@
 	rc = seccomp_arch_add(ctx, seccomp_arch_resolve_name("aarch64"));
 	if (rc != 0)
 		goto out;
+	rc = seccomp_arch_add(ctx, seccomp_arch_resolve_name("e2k"));
+	if (rc != 0)
+		goto out;
 	rc = seccomp_arch_add(ctx, seccomp_arch_resolve_name("mipsel"));
 	if (rc != 0)
 		goto out;
diff -Naur a/src/arch-e2k.c b/src/arch-e2k.c
--- a/src/arch-e2k.c	1970-01-01 03:00:00.000000000 +0300
+++ b/src/arch-e2k.c	2025-02-06 15:38:09.228999861 +0300
@@ -0,0 +1,20 @@
+#include <stdlib.h>
+#include <errno.h>
+#include <linux/audit.h>
+
+#include "arch.h"
+#include "arch-e2k.h"
+#include "syscalls.h"
+
+ARCH_DEF(e2k)
+
+const struct arch_def arch_def_e2k = {
+	.token = SCMP_ARCH_E2K,
+	.token_bpf = AUDIT_ARCH_E2K,
+	.size = ARCH_SIZE_64,
+	.endian = ARCH_ENDIAN_LITTLE,
+	.syscall_resolve_name = e2k_syscall_resolve_name,
+	.syscall_resolve_num = e2k_syscall_resolve_num,
+	.syscall_rewrite = NULL,
+	.rule_add = NULL,
+};
diff -Naur a/src/arch-e2k.h b/src/arch-e2k.h
--- a/src/arch-e2k.h	1970-01-01 03:00:00.000000000 +0300
+++ b/src/arch-e2k.h	2025-02-06 15:37:43.796499435 +0300
@@ -0,0 +1,8 @@
+#ifndef _ARCH_E2K_H
+#define _ARCH_E2K_H
+
+#include "arch.h"
+
+ARCH_DECL(e2k)
+
+#endif
diff -Naur a/src/arch-syscall-dump.c b/src/arch-syscall-dump.c
--- a/src/arch-syscall-dump.c	2025-02-06 15:18:54.102286048 +0300
+++ b/src/arch-syscall-dump.c	2025-02-06 15:39:27.346536901 +0300
@@ -106,6 +107,9 @@
 		case SCMP_ARCH_AARCH64:
 			sys = aarch64_syscall_iterate(iter);
 			break;
+		case SCMP_ARCH_E2K:
+			sys = e2k_syscall_iterate(iter);
+			break;
 		case SCMP_ARCH_MIPS:
 		case SCMP_ARCH_MIPSEL:
 			sys = mips_syscall_iterate(iter);
diff -Naur a/tools/scmp_arch_detect.c b/tools/scmp_arch_detect.c
--- a/tools/scmp_arch_detect.c	2025-02-06 15:22:16.550265033 +0300
+++ b/tools/scmp_arch_detect.c	2025-02-06 15:48:44.357494643 +0300
@@ -81,6 +81,9 @@
 		case SCMP_ARCH_AARCH64:
 			printf("aarch64\n");
 			break;
+		case SCMP_ARCH_E2K:
+			printf("e2k\n");
+			break;
 		case SCMP_ARCH_MIPS:
 			printf("mips\n");
 			break;
diff -Naur a/include/seccomp.h.in b/include/seccomp.h.in
--- a/include/seccomp.h.in	2025-02-06 15:16:50.015846920 +0300
+++ b/include/seccomp.h.in	2025-02-06 15:35:30.725880923 +0300
@@ -151,6 +151,11 @@
 #define SCMP_ARCH_AARCH64	AUDIT_ARCH_AARCH64
 
 /**
+ * The E2K architecture tokens
+ */
+#define SCMP_ARCH_E2K		AUDIT_ARCH_E2K
+
+/**
  * The MIPS architecture tokens
  */
 #ifndef __AUDIT_ARCH_CONVENTION_MIPS64_N32
diff -Naur a/include/seccomp-syscalls.h b/include/seccomp-syscalls.h
--- a/include/seccomp-syscalls.h	2025-02-06 15:16:16.091180034 +0300
+++ b/include/seccomp-syscalls.h	2025-02-06 15:34:33.540755587 +0300
@@ -278,6 +278,14 @@
 #define __PNR_memfd_secret			-10244
 #define __PNR_map_shadow_stack			-10245
 
+#define __PNR_cachestat				-15014
+#define __PNR_fchmodat2				-15015
+#define __PNR_futex_requeue			-15016
+#define __PNR_futex_wait			-15017
+#define __PNR_futex_wake			-15018
+#define __PNR_pread64				-15019
+#define __PNR_pwrite64				-15020
+
 /*
  * libseccomp syscall definitions
  */
@@ -402,7 +410,11 @@
 #define __SNR_cacheflush		__PNR_cacheflush
 #endif
 
+#ifdef __NR_cachestat
 #define __SNR_cachestat			__NR_cachestat
+#else
+#define __SNR_cachestat			__PNR_cachestat
+#endif
 
 #define __SNR_capget			__NR_capget
 
@@ -598,7 +610,11 @@
 
 #define __SNR_fchmodat			__NR_fchmodat
 
+#ifdef __NR_fchmodat2
 #define __SNR_fchmodat2			__NR_fchmodat2
+#else
+#define __SNR_fchmodat2			__PNR_fchmodat2
+#endif
 
 #ifdef __NR_fchown
 #define __SNR_fchown			__NR_fchown
@@ -722,7 +738,11 @@
 
 #define __SNR_futex			__NR_futex
 
+#ifdef __NR_futex_requeue
 #define __SNR_futex_requeue		__NR_futex_requeue
+#else
+#define __SNR_futex_requeue		__PNR_futex_requeue
+#endif
 
 #ifdef __NR_futex_time64
 #define __SNR_futex_time64		__NR_futex_time64
@@ -730,11 +750,19 @@
 #define __SNR_futex_time64		__PNR_futex_time64
 #endif
 
+#ifdef __NR_futex_wait
 #define __SNR_futex_wait		__NR_futex_wait
+#else
+#define __SNR_futex_wait		__PNR_futex_wait
+#endif
 
 #define __SNR_futex_waitv		__NR_futex_waitv
 
+#ifdef __NR_futex_wake
 #define __SNR_futex_wake		__NR_futex_wake
+#else
+#define __SNR_futex_wake		__PNR_futex_wake
+#endif
 
 #ifdef __NR_futimesat
 #define __SNR_futimesat			__NR_futimesat
@@ -1412,7 +1440,11 @@
 
 #define __SNR_prctl			__NR_prctl
 
+#ifdef __NR_pread64
 #define __SNR_pread64			__NR_pread64
+#else
+#define __SNR_pread64			__PNR_pread64
+#endif
 
 #define __SNR_preadv			__NR_preadv
 
@@ -1456,7 +1488,11 @@
 #define __SNR_putpmsg			__PNR_putpmsg
 #endif
 
+#ifdef __NR_pwrite64
 #define __SNR_pwrite64			__NR_pwrite64
+#else
+#define __SNR_pwrite64			__PNR_pwrite64
+#endif
 
 #define __SNR_pwritev			__NR_pwritev
 
diff -Naur a/src/syscalls-e2k.sh b/src/syscalls-e2k.sh
--- a/src/syscalls-e2k.sh	1970-01-01 03:00:00.000000000 +0300
+++ b/src/syscalls-e2k.sh	2025-02-06 15:44:24.564384310 +0300
@@ -0,0 +1,50 @@
+#!/bin/bash
+
+sys_csv="src/syscalls.csv"
+sys_add="src/syscalls_e2k.txt"
+new_csv="src/syscalls_new.csv"
+new_hdr="src/syscalls_hdr.txt"
+
+awk '/^#define __NR/ {
+  # remove duplicate syscall
+  if ($2 == "__NR_madvise1") next;
+  sub("^__NR_","",$2); print $2 " " $3
+}
+# skip tail
+$2 == "NR_syscalls" { exit }
+' /usr/include/asm/unistd.h > $sys_add
+
+echo "$(head -n1 $sys_csv),e2k" > $new_hdr
+
+cat "$sys_csv" | grep '^[^#].*,' | (
+  while IFS= read str; do
+    name=${str%%,*}
+    val="$(grep "^$name " $sys_add | sed "s/^.* //")"
+    test -z "$val" && val="PNR"
+    echo "$str,$val" >> "$new_csv"
+  done
+)
+
+pnr="$(cat "$sys_csv" | grep -m1 '^[^#].*,')"
+pnr="$(echo ${pnr#*,} | sed 's/[0-9]\+/PNR/g')"
+
+cat "$sys_add" | grep '^[^#].* [0-9]\+' | (
+  while IFS= read str; do
+    name=${str%% *}
+    grep -q "^$name," $sys_csv || { \
+      echo $name,$pnr,${str#* } >> "$new_csv"
+      {
+        echo "#define __PNR_$name -(20000+${str#* })"
+        echo "#ifdef __NR_$name"
+        echo "#define __SNR_$name __NR_$name"
+        echo "#else"
+        echo "#define __SNR_$name __PNR_$name"
+        echo "#endif"
+      } >> include/seccomp-syscalls.h
+    }
+  done
+)
+
+mv -f $new_hdr $sys_csv
+sort -g $new_csv >> $sys_csv
+
