--- a/lib/defines.h	2024-06-21 09:00:20.000000000 +0000
+++ b/lib/defines.h	2026-03-09 08:59:58.419330730 +0000
@@ -229,7 +229,7 @@
 /* Maximum length of passwd entry */
 #define PASSWD_ENTRY_MAX_LENGTH 32768
 
-#if (__GNUC__ >= 11) && !defined(__clang__)
+#if (__GNUC__ >= 11) && !defined(__clang__) && !defined(__MCST__)
 # define ATTR_MALLOC(deallocator)  [[gnu::malloc(deallocator)]]
 #else
 # define ATTR_MALLOC(deallocator)
