diff -Naur lame-3.99.5-orig/configure lame-3.99.5/configure
--- lame-3.99.5-orig/configure.in	2012-02-28 22:54:37.000000000 +0400
+++ lame-3.99.5/configure.in	2016-01-01 13:41:19.713625559 +0300
@@ -680,7 +680,7 @@
 dnl ### processor specific options ###
 WITH_VECTOR=no
 case $host_cpu in
-x86_64|amd64)
+x86_64|amd64|e2k)
 	CPUTYPE="no"
 	if test $ac_cv_header_xmmintrin_h = yes ; then
 		WITH_XMM=yes
