diff --git a/sysdeps/e2k/e2k128/Makefile b/sysdeps/e2k/e2k128/Makefile
index a0009fb..172fdae 100644
--- a/sysdeps/e2k/e2k128/Makefile
+++ b/sysdeps/e2k/e2k128/Makefile
@@ -1,4 +1,4 @@
-e2k128-cflags =
+e2k128-cflags = -mptr128
 
 ifeq (yes,$(compile-with-lcc))
 e2k128-cflags += -Wno-void-pointer-to-function-pointer-cast
diff --git a/sysdeps/e2k/e2k32/Makefile b/sysdeps/e2k/e2k32/Makefile
index d303a12..1c070b5 100644
--- a/sysdeps/e2k/e2k32/Makefile
+++ b/sysdeps/e2k/e2k32/Makefile
@@ -1,4 +1,4 @@
-e2k32-cflags =
+e2k32-cflags = -mptr32
 
 CC += $(e2k32-cflags)
 CXX += $(e2k32-cflags)
diff --git a/sysdeps/e2k/e2k64/Makefile b/sysdeps/e2k/e2k64/Makefile
index 35593f9..5cdcb82 100644
--- a/sysdeps/e2k/e2k64/Makefile
+++ b/sysdeps/e2k/e2k64/Makefile
@@ -1,4 +1,4 @@
-e2k64-cflags =
+e2k64-cflags = -mptr64
 
 CC += $(e2k64-cflags)
 CXX += $(e2k64-cflags)
