diff --git a/sysdeps/e2k/preconfigure b/sysdeps/e2k/preconfigure
index 81150712ee..4c71732cce 100644
--- a/sysdeps/e2k/preconfigure
+++ b/sysdeps/e2k/preconfigure
@@ -1,9 +1,9 @@
 # preconfigure fragment for e2k.
 
 case "$machine" in
-e2k | e2k32)
+e2k32)
 		base_machine=e2k machine=e2k/e2k32 ;;
-e2k64)
+e2k | e2k64)
 		base_machine=e2k machine=e2k/e2k64 ;;
 e2k128)
 		base_machine=e2k machine=e2k/e2k128 ;;
